### PR TITLE
Added medical datasets

### PR DESCRIPTION
**Pull request by nomisto**

### CODE REVIEW

**Comment by matthias-samwald** [comment]
@nomisto Could you share the resulting JSON files?

**Comment by nomisto** [comment]
Sure, you can download it [here](https://we.tl/t-fQO8PKczFY). You can also generate the json yourself by running:
```python
from cot import Collection
collection = Collection(["med_qa", "medmc_qa", "pubmed_qa"])
collection.dump("dump.json")
```

The following examples should now have CoTs from Lievin:
500 random in pubmedqa/train
all in medqa/test
all in medmcqa/validation

**Comment by matthias-samwald** [comment]
Spotted in the first item of med_qa/test: One of the answers is marked as with "correct_answer": false, while all other are marked as true. However, it seems like all the answers as identical strings.

```
test": [
            {
                "id": "0",
                "question_id": "0",
                "document_id": "0",
                "question": "A junior orthopaedic surgery resident is completing a carpal tunnel repair with the department chairman as the attending physician. During the case, the resident inadvertently cuts a flexor tendon. The tendon is repaired without complication. The attending tells the resident that the patient will do fine, and there is no need to report this minor complication that will not harm the patient, as he does not want to make the patient worry unnecessarily. He tells the resident to leave this complication out of the operative report. Which of the following is the correct next action for the resident to take?",
                "type": "multiplechoice",
                "cot_type": "list",
                "choices": [
                    "Disclose the error to the patient but leave it out of the operative report",
                    "Disclose the error to the patient and put it in the operative report",
                    "Tell the attending that he cannot fail to disclose this mistake",
                    "Report the physician to the ethics committee",
                    "Refuse to dictate the operative report"
                ],
                "context": "",
                "cot": [],
                "answer": [
                    "Tell the attending that he cannot fail to disclose this mistake"
                ],
                "generated_cot": [
                    {
                        "templates_version": "0.01",
                        "instruction": null,
                        "cot-trigger": "kojima-01",
                        "answers": [
                            {
                                "answer-extraction": "kojima-A-D",
                                "answer": "Tell the attending that he cannot fail to disclose this mistake",
                                "correct_answer": false
                            }
                        ],
                        "cot": "Let's think step by step about what the resident should do. \n\nThe first step is to talk to the attending physician. The attending is the one who made the decision not to disclose the error, so the resident should talk to him or her about why they made that decision and see if they can come to a consensus about what to do. \n\nIf the attending is adamant about not disclosing the error, then the resident has a few options. The resident could go to the ethics committee, but that may not be the best option. It's possible that the attending is acting in the best interest of the patient and that disclosing the error could cause more harm than good. \n\nThe resident could also refuse to dictate the operative report, but that may not be the best option either. The operative report is a medical record and should be accurate. Refusing to dictate the report could cause problems down the line if the error is discovered. \n\nThe best option for the resident is probably to disclose the error to the patient. The patient has a right to know what happened during their surgery, and the resident should make sure that they are informed.",
                        "author": "lievin",
                        "date": null,
                        "model": "davinci-002",
                        "comment": "",
                        "annotation": []
                    },
                    {
                        "templates_version": "0.01",
                        "instruction": null,
                        "cot-trigger": "lievin-10",
                        "answers": [
                            {
                                "answer-extraction": "kojima-A-D",
                                "answer": "Tell the attending that he cannot fail to disclose this mistake",
                                "correct_answer": true
                            }
                        ],
                        "cot": "Let's think step by step like a medical expert. \n\nThe first thing the resident should do is to tell the attending that he cannot fail to disclose this mistake. This is the most important thing to do in order to protect the patient's safety. \n\nThe second thing the resident should do is to report the physician to the ethics committee. This is the best way to make sure that the attending does not make this mistake again in the future.",
                        "author": "lievin",
                        "date": null,
                        "model": "davinci-002",
                        "comment": "",
                        "annotation": []
                    },
                    {
                        "templates_version": "0.01",
                        "instruction": null,
                        "cot-trigger": "lievin-01",
                        "answers": [
                            {
                                "answer-extraction": "kojima-A-D",
                                "answer": "Tell the attending that he cannot fail to disclose this mistake",
                                "correct_answer": false
                            }
                        ],
                        "cot": "Let's derive the differential diagnosis step by step:\n\n1) The attending physician tells the resident to leave this complication out of the operative report.\n\n2) The resident refuses to dictate the operative report.\n\n3) The resident tells the attending that he cannot fail to disclose this mistake.\n\n4) The resident discloses the error to the patient and puts it in the operative report.\n\nThe most likely diagnosis is 4) The resident discloses the error to the patient and puts it in the operative report.",
                        "author": "lievin",
                        "date": null,
                        "model": "davinci-002",
                        "comment": "",
                        "annotation": []
                    },
                    {
                        "templates_version": "0.01",
                        "instruction": null,
                        "cot-trigger": "lievin-02",
                        "answers": [
                            {
                                "answer-extraction": "kojima-A-D",
                                "answer": "Tell the attending that he cannot fail to disclose this mistake",
                                "correct_answer": true
                            }
                        ],
                        "cot": "Let's use step by step inductive reasoning, given the medical nature of the question. \n\nThe first thing the resident should do is consult with the attending physician to get his or her opinion on what happened and what should be done next. \n\nThe second thing the resident should do is make sure that the patient is informed of the error and that the error is documented in the operative report. \n\nThe third thing the resident should do is report the physician to the ethics committee.",
                        "author": "lievin",
                        "date": null,
                        "model": "davinci-002",
                        "comment": "",
                        "annotation": []
                    },
                    {
                        "templates_version": "0.01",
                        "instruction": null,
                        "cot-trigger": "lievin-03",
                        "answers": [
                            {
                                "answer-extraction": "kojima-A-D",
                                "answer": "Tell the attending that he cannot fail to disclose this mistake",
                                "correct_answer": true
                            }
                        ],
                        "cot": "Let's differentiate using step by step reasoning like a medical expert. \n\nThe first thing the resident should do is to tell the attending that he cannot fail to disclose this mistake. This is the correct next action for the resident to take.",
                        "author": "lievin",
                        "date": null,
                        "model": "davinci-002",
                        "comment": "",
                        "annotation": []
                    }
                ],
                "feedback": []
```

**Comment by matthias-samwald** [comment]
Looking at further items in med_qa, this pattern seems to be repeating. All of the generated answers I checked are the same for each question (which is maybe an error in itself), but some of the answers are evaluated as correct and some as incorrect.

**Comment by nomisto** [comment]
Ah, thanks for spotting that! I confused the field "answer" in Lievins data to be the predicted answer, however it is the gold answer, thus it is the same on every CoT

**Comment by matthias-samwald** [comment]
For medmc_qa, several confusing things: The gold-standard CoT is filled, but I don't know where the text is coming from. At first I thougt it would be the context that they auto-generated from Wikipedia/PubMed, but that is not it either. I cannot find the text in https://vlievin.github.io/medical-reasoning/samples/medmcqa-wo-context.html

In general, I cannot find the some of the IDs (like the one of the examples below) in https://vlievin.github.io/medical-reasoning/samples/medmcqa-wo-context.html

```
medmc_qa": {
        "train": [
            {
                "id": "e9ad821a-c438-4965-9f77-760819dfa155",
                "question_id": "e9ad821a-c438-4965-9f77-760819dfa155",
                "document_id": "e9ad821a-c438-4965-9f77-760819dfa155",
                "question": "Chronic urethral obstruction due to benign prismatic hyperplasia can lead to the following change in kidney parenchyma",
                "type": "multiplechoice",
                "cot_type": "list",
                "choices": [
                    "Hyperplasia",
                    "Hyperophy",
                    "Atrophy",
                    "Dyplasia"
                ],
                "context": "",
                "cot": [
                    "Chronic urethral obstruction because of urinary calculi, prostatic hyperophy, tumors, normal pregnancy, tumors, uterine prolapse or functional disorders cause hydronephrosis which by definition is used to describe dilatation of renal pelvis and calculus associated with progressive atrophy of the kidney due to obstruction to the outflow of urine Refer Robbins 7yh/9,1012,9/e. P950"
                ],
                "answer": [
                    "Atrophy"
                ],
                "generated_cot": [],
                "feedback": []
            },
            {
                "id": "e3d3c4e1-4fb2-45e7-9f88-247cc8f373b3",
                "question_id": "e3d3c4e1-4fb2-45e7-9f88-247cc8f373b3",
                "document_id": "e3d3c4e1-4fb2-45e7-9f88-247cc8f373b3",
                "question": "Which vitamin is supplied from only animal source:",
                "type": "multiplechoice",
                "cot_type": "list",
                "choices": [
                    "Vitamin C",
                    "Vitamin B7",
                    "Vitamin B12",
                    "Vitamin D"
                ],
                "context": "",
                "cot": [
                    "Ans. (c) Vitamin B12 Ref: Harrison's 19th ed. P 640* Vitamin B12 (Cobalamin) is synthesized solely by microorganisms.* In humans, the only source for humans is food of animal origin, e.g., meat, fish, and dairy products.* Vegetables, fruits, and other foods of nonanimal origin doesn't contain Vitamin B12 .* Daily requirements of vitamin Bp is about 1-3 pg. Body stores are of the order of 2-3 mg, sufficient for 3-4 years if supplies are completely cut off."
                ],
                "answer": [
                    "Vitamin B12"
                ],
                "generated_cot": [],
                "feedback": []
            },
```

**Comment by nomisto** [comment]
The gold CoTs is from the field `exp` (Expert's explanation of the answer) https://github.com/medmcqa/medmcqa#data-fields in the original dataset (medmcqa)[https://github.com/medmcqa/medmcqa] (These are only present in the training dataset) and I decided to put them into as gold cot if that makes sense. Lievin does not include them in their data.

Regarding the IDs: Lievin only uses the validation set of medmcqa (for which they do CoT prompting). Our dataloader has the full original medmcqa and adds Lievins CoTs to examples of the validation set.

**Comment by nomisto** [comment]
Ah, yes I've made a comment about this here:

https://github.com/OpenBioLink/ThoughtSource/blob/aaef629daeb49832d615d20f5963be56c80144b4/libs/cot/cot/utils/utils.py#L244-L249

F.e. for the first example if I would do this the only thing remaining would be "about what the resident should do. \n\nThe first step ..."

**Comment by matthias-samwald** [comment]
Note for future reference: The CoT trigger prompts that Lievin et al. used did not end with a full stop, i.e. they used "Let's think step by step" instead of "Let's think step by step."
This means that the model sometimes/often continued the sentence. This also means that we cannot easily remove the CoT triggers from the CoTs in these datasets.

**Review by matthias-samwald — APPROVED** [review]
